coderd/schedule/autostop.go

```diff
@@ -0,0 +1,234 @@
+// past this value, so it denotes the absolute deadline that the workspace build
+// must be stopped by. MaxDeadline is calculated using the template's "restart
+// requirement" settings and the user's "quiet hours" settings to pick a time
+// outside of working hours.
+//
+// Deadline is a cost saving measure, while max deadline is a
+// compliance/updating measure.
+func CalculateAutostop(ctx context.Context, params CalculateAutostopParams) (AutostopTime, error) {
+	var (
+		db        = params.Database
+		workspace = params.Workspace
+		now       = params.Now
+
+		autostop AutostopTime
+	)
+
+	if workspace.Ttl.Valid {
+		// When the workspace is made it copies the template's TTL, and the user
+		// can unset it to disable it (unless the template has
+		// UserAutoStopEnabled set to false, see below).
+		autostop.Deadline = now.Add(time.Duration(workspace.Ttl.Int64))
+	}
+
+	templateSchedule, err := params.TemplateScheduleStore.GetTemplateScheduleOptions(ctx, db, workspace.TemplateID)
+	if err != nil {
+		return autostop, xerrors.Errorf("get template schedule options: %w", err)
+	}
+	if !templateSchedule.UserAutostopEnabled {
+		// The user is not permitted to set their own TTL, so use the template
+		// default.
+		autostop.Deadline = time.Time{}
+		if templateSchedule.DefaultTTL > 0 {
+			autostop.Deadline = now.Add(templateSchedule.DefaultTTL)
+		}
+	}
+
+	if templateSchedule.RestartRequirement.DaysOfWeek != 0 {
+		// The template has a restart requirement, so determine the max deadline
+		// of this workspace build.
+
+		// First, get the user's quiet hours schedule (this will return the
+		// default if the user has not set their own schedule).
+		userQuietHoursSchedule, err := params.UserQuietHoursScheduleStore.GetUserQuietHoursScheduleOptions(ctx, db, workspace.OwnerID)
+		if err != nil {
+			return autostop, xerrors.Errorf("get user quiet hours schedule options: %w", err)
+		}
+
+		// If the schedule is nil, that means the deployment isn't entitled to
+		// use quiet hours or the default schedule has not been set. In this
+		// case, do not set a max deadline on the workspace.
+		if userQuietHoursSchedule.Schedule != nil {
+			loc := userQuietHoursSchedule.Schedule.Location()
+			now := now.In(loc)
+			// Add the leeway here so we avoid checking today's quiet hours if
+			// the workspace was started <1h before midnight.
+			startOfStopDay := truncateMidnight(now.Add(restartRequirementLeeway))
+
+			// If the template schedule wants to only restart on n-th weeks then
+			// change the startOfDay to be the Monday of the next applicable
+			// week.
+			if templateSchedule.RestartRequirement.Weeks > 1 {
+				epoch := TemplateRestartRequirementEpoch(loc)
+				if startOfStopDay.Before(epoch) {
+					return autostop, xerrors.New("coder server system clock is incorrect, cannot calculate template restart requirement")
+				}
+				since := startOfStopDay.Sub(epoch)
+				weeksSinceEpoch := int64(since.Hours() / (24 * 7))
```

Comment:
This algorithm might be too naive??? More research/testing required to ensure this generates correct dates

Reply:
Redid the algorithm completely since this algorithm fails if DST has passed once already during the year. The new algorithm is good in a bunch of timezones for the next 100 years (and this is tested).